### PR TITLE
Google My Business: Add line charts

### PR DESCRIPTION
**Pull request by cbauerman**
This pull request introduces a brand new line chart component, and updates the Google My Business `Stats` tab to use it:
 
screenshot
  
#### Testing instructions
 
1. Run `git checkout try/gmb-stats-charts` and start your server, or open a [live branch](https://calypso.live/?branch=try/gmb-stats-charts)
2. Open the [`Sharing` page](http://calypso.localhost:3000/sharing)
3. Connect to Google My Business
4. Navigate to the [`Stats` page](http://calypso.localhost:3000/stats)
5. Click the `Google My Business` link in the header cake
6. Assert that the `Stats` tab looks like the screenshot above
7. Assert that changing an interval updates the corresponding chart

#### Reviews
 
- [ ] Code
- [ ] Product

### CODE REVIEW

**Comment by stephanethomas** [comment]
This looks promising. I've rebased, and addressed all my comments except the one about using new React lifecycle methods.

I did look into it, but this actually raised more questions about how this `D3Base` component is designed, and how it should react to data changes. As mentioned during our hangout, I kind of feel we have two options: either we carry on with the current approach, and at some point remove the original component from the WooCommerce extension - or we take the time to rethink it, and define a new Calypso standard. However the latter scares me because of the additional effort required, and the fact that this WooCommerce component was probably designed this way for very good reasons.

Anyway, these line charts looks good enough for now. I tried to customize them this afternoon but didn't go really far because of how complex this task was. I think the only thing that is missing is the legend. We don't really have to make items from this legend clickable like in the Google My Business dashboard. I do think though that it would be a good idea to have a separate component similar to `PieChartLegend` that would give us lots of flexibility in term of layout.

**Comment by stephanethomas** [comment]
I've rebased, fixed a few things, and updated the page so line and pie charts can live together. This is still not perfect though, I've noticed the following issues:

* Plenty of React warnings because `data` is null
* Tooltips don't work very well when the window is resized
* Dots are lost when intervals change

We can address them in another pull request though.

File: client/my-sites/google-my-business/stats/chart.js [hunk 1]
```diff
@@ -24,70 +25,97 @@ import { getGoogleMyBusinessStats } from 'state/selectors';
...
+function transformData( props ) {
+	const { data } = props;
+
+	if ( ! data ) {
+		return data;
+	}
+
+	const aggregation = getAggregation( props );
+
+	if ( aggregation === 'total' ) {
```
Comment (Tug) [inline client/my-sites/google-my-business/stats/chart.js]: Note: this is modified in https://github.com/Automattic/wp-calypso/pull/24236 as well

**Comment by Tug** [comment]
This PR contains some refactoring that is necessary for other PRs so we should aim for a merge and come back to it to fix the following issues I noticed:
- Add a legend
- Axis: show less units for more clarity. No need to display x,5 steps and maybe use a bigger text?
- Handle `null` values in `getTooltipContent()`
- Use a placeholder when chart is drawing, otherwise the whole page layout can change as the chart disapears (I think I preferred the previous behavior which was keeping the old chart until the new one is loaded)
- Refine chart dimensions: I think we should reduce the height significantly when the page shows 1 column as it doesn't bring much value having it take the whole page imo
- Show points on the curve for months and year as well, maybe aggregate the data if necessary
- Add units to the axis (date and number of views for instance).
- Show a notice when no data (values at 0). See `/stats/day/:site` which displays "No activity this period"
- `D3Base` is rather complex at the moment, we should investigate a way to simplify it or write a good documentation with examples. Also `render` is triggered about 10 times on first load which is probably unnecessary, need to investigate.
- Find a way to group the chart components together, they can really be separated, especially when the page is displayed on 2 columns. We could use that opportunity to group the interval selector into 1 component, so all charts update at the same time when we change the interval.
- Remove the alert "All data on this page is a placeholder used for development only!"
- Refine the CSS

File: client/components/d3-base/index.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	 * case (in order to indicate that the chart should be drawn again).
+	 */
+	static getDerivedStateFromProps( nextProps, prevState ) {
+		let state = {};
+
+		if ( nextProps.data !== prevState.data ) {
+			state = { ...state, data: nextProps.data };
+		}
+
+		if ( nextProps.drawChart !== prevState.drawChart ) {
```
Comment (Tug) [inline client/components/d3-base/index.js]: `drawChart` and `getParams` should not be allowed to change I think. We should spend a bit more time on D3Base when we come back to it I think

**Comment by stephanethomas** [comment]
> D3Base is rather complex at the moment, we should investigate a way to simplify it or write a good documentation with examples. Also render is triggered about 10 times on first load which is probably unnecessary, need to investigate.

`render` is only called twice for each chart, and that's because of the new React lifecycle methods API: i.e. using `getDerivedStateFromProps()` and the fact that we should only perform DOM operations in `componentDidMount()` and `componentDidUpdate()`.

> Find a way to group the chart components together, they can really be separated, especially when the page is displayed on 2 columns. We could use that opportunity to group the interval selector into 1 component, so all charts update at the same time when we change the interval.

I don't see that as a problem, but let's bring that up with Courtney during her design/copy review.

**Comment by Tug** [comment]
> render is only called twice for each chart ...

Indeed, it looks like your update fixed that!

**Comment by Tug** [comment]
Another thing we should update is `GoogleMyBusinessStatsChart`, in particular how it's used in `GoogleMyBusinessStats`:
- the `chartType` property is not consistent, only one has it defined
- `renderTooltipForDatanum` is only relevant for `LineChart`, we can remove `renderTooltipForDatanum` but we need a way pass metadata about what's being displayed are they views, searches, etc?

**Comment by Tug** [comment]
> Refine the CSS

If you look at the page at the moment, the margin and the font size is not consistent, the legend of the pie chart is way too big, also the component height is too large as I mentioned, I think it needs a bit of refining so it looks better.

@stephanethomas LGTM 👍 Could you regenerate `npm-shrinkwrap.json` before you merge!

**Comment by stephanethomas** [comment]
I've just rebased to address conflicts in `npm-shrinkwrap.json`.

**Pull request merged**